app/validation/validator.py

```diff
@@ -165,13 +165,23 @@ def validate_schema_routing_rule_routes_to_valid_target(self, dict_list, goto_ke
                 errors.append(self._error_message(invalid_block_error))
         return errors
 
-    @staticmethod
-    def validate_schema_routing_rule_dependent_on_valid_answer(rule, answer_ids_with_group_id, block_or_group):
+    def validate_routing_rule(self, rule, all_groups, answer_ids_with_group_id, block_or_group):
         errors = []
 
-        rule = rule.get('goto') or rule.get('repeat')
-        if 'when' in rule:
-            when_errors = Validator.validate_when_rule(rule['when'], answer_ids_with_group_id, block_or_group['id'])
+        when = None
+        if 'goto' in rule:
+            when = rule['goto'].get('when')
+        elif 'repeat' in rule:
```

Comment:
Something about this block gives me the feeling it wont be very readable in a few months. Perhaps a method for `validate_repeat_over_group_ids` could be added, then you can test for that special case and return early from this method?